linux/LinuxMachine.c

```diff
@@ -820,3 +824,108 @@ bool Machine_isCPUonline(const Machine* super, unsigned int id) {
    assert(id < super->existingCPUs);
    return this->cpuData[id + 1].online;
 }
+
+static int humanTimeUnit(char* buffer, size_t size, unsigned long long int value) {
+
+   if (value < 1000)
+      return xSnprintf(buffer, size, "%3lluns", value);
+
+   if (value < 10000)
+      return xSnprintf(buffer, size, "%1llu.%1lluus", value / 1000, (value % 1000) / 100);
+
+   value /= 1000;
+
+   if (value < 1000)
+      return xSnprintf(buffer, size, "%3lluus", value);
+
+   if (value < 10000)
+      return xSnprintf(buffer, size, "%1llu.%1llums", value / 1000, (value % 1000) / 100);
+
+   value /= 1000;
+
+   if (value < 1000)
+      return xSnprintf(buffer, size, "%3llums", value);
+
+   if (value < 10000)
+      return xSnprintf(buffer, size, "%1llu.%1llus", value / 1000, (value % 1000) / 100);
+
+   value /= 1000;
+
+   if (value < 600)
+      return xSnprintf(buffer, size, "%3llus", value);
+
+   value /= 60;
+
+   if (value < 600)
+      return xSnprintf(buffer, size, "%3llum", value);
+
+   value /= 60;
+
+   if (value < 96)
+      return xSnprintf(buffer, size, "%3lluh", value);
+
+   value /= 24;
+
+   return xSnprintf(buffer, size, "%3llud", value);
+}
+
+double Machine_getGpuUsage(const Machine* super, double values[]) {
+   unsigned int i;
+   uint64_t monotonictimeDelta;
+   const LinuxMachine* this = (const LinuxMachine*) super;
+   GPUEngineData *gpuEngineData;
+
+   assert(ARRAYSIZE(GPUMeter_engineData) + 1 == ARRAYSIZE(GPUMeter_attributes));
+
+	totalGPUTimeDiff = saturatingSub(this->curGpuTime, this->prevGpuTime);
+   monotonictimeDelta = super->monotonicMs - super->prevMonotonicMs;
+
+   prevResidueTime = curResidueTime;
+   curResidueTime = this->curGpuTime;
+
+   for (gpuEngineData = this->gpuEngineData, i = 0; gpuEngineData && i < ARRAYSIZE(GPUMeter_engineData); gpuEngineData = gpuEngineData->next, i++) {
+      GPUMeter_engineData[i].key      = gpuEngineData->key;
+      GPUMeter_engineData[i].timeDiff = saturatingSub(gpuEngineData->curTime, gpuEngineData->prevTime);
+
+      curResidueTime = saturatingSub(curResidueTime, gpuEngineData->curTime);
+
+      values[i] = 100.0 * GPUMeter_engineData[i].timeDiff / (1000 * 1000) / monotonictimeDelta;
+   }
+
+   values[ARRAYSIZE(GPUMeter_engineData)] = 100.0 * saturatingSub(curResidueTime, prevResidueTime) / (1000 * 1000) / monotonictimeDelta;
+
+   return 100.0 * totalGPUTimeDiff / (1000 * 1000) / monotonictimeDelta;
+}
+
+void Machine_GPUMeterDisplay(const Object* cast, RichString* out, double totalUsage) {
```

Comment:
I can't see how it make sense for having this function Linux-only. (This function was formerly named `GPUMeter_display`, IIRC.) The text for displaying the GPU stat should be platform independent, no?

Reply:
It absolutely should be platform-independent …